### DISCUSSION TITLE
读取非初始化目录下的文件时，真正读取的文件名有问题，这是Android遗留的问题，为什么鸿蒙Next上也有此问题？

### DISCUSSION THREAD

**Discussion by peerless2012** [Tencent/MMKV#1378] [Q&A] [answered]
最近正在做Android数据迁移到鸿蒙，发现读取非初始化目录下的mmkv文件时，指定读取的文件名会变成另外一个字符串。查询后发现是[Android历史bug](MMKV::MMKV(const string &mmapID, int size, MMKVMode mode, string *cryptKey, string *rootPath, size_t expectedCapacity))，为什么会出现在鸿蒙上呢？导致我在鸿蒙上也只能用`backedUpMMKVWithID`来读取。

**Comment by lingol** [comment]
It has nothing to do with HarmonyOS NEXT. This is a **historical bug in the Android platform** dating from the very early versions of MMKV. 
For the best practice of HarmonyOS NEXT data migration, call `MMKV.restoreAllFromDirectory()` and read the restored results. https://github.com/Tencent/MMKV/wiki/ohos_setup#backup--restore

**Comment by peerless2012** [comment]
但是，实际上并不存在备份这个操作，系统升级切换到Next的时候，系统只会把沙盒文件拷贝到备份恢复目录，然后做迁移。现在的情况是，我们的mmkv文件都在默认目录下，所以文件名都是正确的。但是在迁移时，读取我只能使用`backedUpMMKVWithID`，因为使用`mmkvWithID`, 会去查找id的文件，自然是找不到的。这个困扰了我好久，通过日志才发现根源。

**Reply by lingol** [reply]
You are not getting it. I mean using that method to do the **immigration job** you described.

**Reply by peerless2012** [reply]
从Android升到鸿蒙NEXT，没有一个地方能让Android的app调用backup方法。

**Reply by lingol** [reply, marked as answer]
OK, I will try to explain it again. When you "做迁移" from the "备份恢复目录", you 
1. calc the root directory of MMKV inside the "备份恢复目录", 
2. call `MMKV.restoreAllFromDirectory("备份恢复目录/xxxx/mmkv")`
3. finally, start using MMKV from your default directory in your new HarmonyOS NEXT App.

Understand?

**Reply by peerless2012** [reply]
明白了。

**Comment by peerless2012** [comment]
并且我看了看代码，这个id错误这个问题，应该是仅限于Android平台，所以鸿蒙Next是否也不应该考虑这个问题？

**Reply by lingol** [reply]
You are also not getting it. The so-called Harmony OS 4 and before are just Android. At least it's the same for MMKV.

**Reply by peerless2012** [reply]
但是鸿蒙NEXT应该是一个全新的平台啊，跟Android和iOS是平级的。

**Reply by lingol** [reply]
The data your old App was running was an Android App, so the filenames of MMKV is in the format due to the Android bug. Now you have migrated to HarmonyOS NEXT, the **filename should stay the same** otherwise **MMKV won't find them**. 

That means HarmonyOS NEXT will have to stick to the same bug.

**Reply by peerless2012** [reply]
我明白了。

**Comment by peerless2012** [comment]
对于不考虑迁移，或者很少数据需要迁移的应用，鸿蒙NEXT是一个新机会，摒弃Android当时的bug，如果能兼顾两者就好了。

**Reply by lingol** [reply]
No, it will be more complicated because now we have to **support two modes of naming** for there's **no way to ignore** data migration from the old Android-compatible system.

**Reply by wangyatong** [reply]
建议直接忽略它, 要做适当的取舍

**Comment by lingol** [comment]
For all those who suffer from **incorrect file names** in Android/OHOS, we are planning a new feature called [NameSpace](https://github.com/Tencent/MMKV/tree/dev_namespace) and will solve this problem once and for all. Under this feature, a previously incorrect file name will be migrated to the correct file name under a custom root path. 

However, this also means that once upgraded, the file will no longer be accessed by older versions of MMKV/your App, which will be a problem if your team member or your users do some downgrade operation. To address this issue, we have provided a forward support version of MMKV that supports reading upgraded/correct file name, and it's released last Friday as [v2.0.2](https://github.com/Tencent/MMKV/releases/tag/v2.0.2).

So please upgrade to v2.0.2 and prepare for this shiny new feature **NameSpace**.
